server/src/main/java/org/apache/kafka/server/ApiVersionManager.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server;
+
+import org.apache.kafka.common.message.ApiMessageType;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.requests.ApiVersionsResponse;
+import org.apache.kafka.network.metrics.RequestChannelMetrics;
+import org.apache.kafka.server.common.FinalizedFeatures;
+
+import java.util.Set;
+
+/**
+ * ApiVersionManagers are used to define the APIs supported by servers
+ */
+public interface ApiVersionManager {
+
+    /**
+     * Whether to mark unstable API versions as enabled
+     * @return true if unstable API versions are enabled, otherwise false
+     */
+    boolean enableUnstableLastVersion();
+
+    /**
+     * The listener type
+     * @return Broker or Controller depending on the server's role
+     */
+    ApiMessageType.ListenerType listenerType();
+
+    /**
+     * The enabled APIs
+     * @return A Set of ApiKeys
+     */
+    Set<ApiKeys> enabledApis();
```

Comment:
Both implementations rely on `ApiKeys.apisForListener(listenerType())`, and this method is only used by `newRequestMetrics`. Hence, maybe we can remove it and then rewrite `newRequestMetrics`
```java
    default RequestChannelMetrics newRequestMetrics() {
        return new RequestChannelMetrics(ApiKeys.apisForListener(listenerType()));
    }
```